### ISSUE TITLE
test cache free of non-cache memory

### ISSUE THREAD

**Issue by goatshriek** [goatshriek/stumpless#441]
Stumpless uses a simple slab cache implementation in `src/cache.c` to consolidate memory allocations for common structures like entries. [Test coverage](https://app.codecov.io/gh/goatshriek/stumpless/blob/latest/src%2Fcache.c) of this cache is fairly complete, but still has some gaps.

One such gap is what happens when free is called on an object that is not owned by the cache. The mutex for the cache should be unlocked once this routine is finished, but currently this is an untested behavior. You will add a test for this case.

## General Approach
There are a few details left out of the following approach, for you to fill in as you encounter them. If you find you need help, please ask here or on the [project gitter](https://gitter.im/stumpless/community) and someone can help you get past the stumbling block.

First, familiarize yourself with the `cache_free` implementation in `src/cache.c`. The last line of this function after the for loop is currently uncovered by tests, and is only hit when the entire cache is searched but the object that has been freed is not found.

Next, devise a test for the freeing of an entry structure that is not in the cache. One way to do this is to add a test to `test/function/entry.cpp` that constructs a new entry normally, allocates a new entry struct dynamically and uses `memcpy` to copy the normal entry, and then calls the entry destructor on the manually allocated one. Remember that you will need to manually free the dynamic structure memory as well, since the `cache_free` call will not. Look at the other destructor tests and the `create_empty_entry` helper function for inspiration for this test.

Make sure that you have full coverage on `cache_free` after this test. See the [test documentation](https://github.com/goatshriek/stumpless/blob/latest/docs/test.md#test-coverage) for details on checking coverage.

Finally, refactor `cache_free` after this to simplify the function to having only one exit pointer. This should be as simple as changing the unlock and return inside the for loop to a break statement. Do this _after_ you have written your test, so that you are sure that the refactored code still works!

**Comment by twingoof** [comment]
Hello, I'll try to fix this and hopefully come with a PR soon (first time contributing I'll do my best)

**Comment by DevTigerart** [comment]
Hi,
If not already addressed I would be happy to work on this.
Best regards

**Comment by goatshriek** [comment]
You may work this one if you would like, but I prefer that contributors only hold one good first issue at a time to avoid locking out other would-be contributors. Do you still intend to work #183? I saw pull request #492, but it seems to have been closed.

**Comment by DevTigerart** [comment]
Oh okay I completely understand !
Then forget I asked, I was just planning for next work after #183 is completed. 
Regarding the PR #492, I closed it because I had issues with commits not being signed (I didn't know about that). I was planning to create another branch with commits being signed properly and then open a new PR so that everything is cleaned.
Is that okay ?

**Comment by goatshriek** [comment]
No worries! I definitely recommend setting up commit signing, but it isn't a strict requirement right now. You don't need to worry too much about commit atomicity either, as I squash PRs on merge, and force pushing ( especially after reviews) makes the changes harder to follow. But good commit discipline is also important to practice, so whichever path you choose just take your time and let me know if you run into issues.

**Comment by rs0125** [comment]
Hey @goatshriek.
I noticed this issue hasn’t seen activity in a while. If it’s not currently assigned or being worked on, I’d be happy to take it up and work on a solution.

Thanks!

**Comment by goatshriek** [comment]
Sure thing! I will assign it to you so that others can see it is in work.

**Comment by rs0125** [comment]
Hi @goatshriek ! I’ve added a test to cover the cache_free behavior when freeing an entry not owned by the slab cache (to hit the final `config_unlock_mutex()` line).

To simulate this, I created a `stumpless_entry` using `create_empty_entry()`, then allocated a new struct and used `memcpy` to clone the original. I then passed the original and then this clone to `stumpless_destroy_entry_only()`.

However, this triggers a double-free under Valgrind, since both the original and the copied struct point to the same internal strings (like message, app_name, etc.), and both are being freed.

I considered calling `cache_free(entry_cache, ptr)` directly, but doing so in C++ leads to compile errors due to _Atomic usage in stdatomic.h. 

What approach would you suggest to execute this cleanly without any double free issues?

Thanks in advance for the guidance!

**Comment by goatshriek** [comment]
Yes, `cache_free` isn't exposed to the public API, so you won't be able to call it in tests without extra work. Since the call to `stumpless_free_all` will free all cache memory anyway, you can probably get away with simply freeing the manual structure via `free` after the test, and relying on the free all to clean up the rest. If this still doesn't work out, you can create a draft pull request and I'll take a closer look.

**Comment by rs0125** [comment]
@goatshriek the draft PR has been raised. Kindly have a look and let me know if there's a better approach for this problem.

Thank you!
